src/main/scala/firrtl/backends/experimental/smt/random/InvalidToRandomPass.scala

```diff
@@ -0,0 +1,128 @@
+    require(opts.size < 2, s"Multiple options: $opts")
+    val opt = opts.headOption.getOrElse(InvalidToRandomOptions())
+
+    // quick exit if we just want to skip this pass
+    if (!opt.randomizeDivisionByZero && !opt.randomizeInvalidSignals) {
+      state
+    } else {
+      val c = state.circuit.mapModule(onModule(_, opt))
+      state.copy(circuit = c)
+    }
+  }
+
+  private def onModule(m: DefModule, opt: InvalidToRandomOptions): DefModule = m match {
+    case d: DescribedMod =>
+      throw new RuntimeException(s"CompilerError: Unexpected internal node: ${d.serialize}")
+    case e:   ExtModule => e
+    case mod: Module =>
+      val namespace = Namespace(mod)
+      mod.mapStmt(onStmt(namespace, opt, _))
+  }
+
+  private def onStmt(namespace: Namespace, opt: InvalidToRandomOptions, s: Statement): Statement = s match {
+    case IsInvalid(info, loc: RefLikeExpression) if opt.randomizeInvalidSignals =>
+      val name = namespace.newName(loc.serialize.replace('.', '_') + "_invalid")
+      val rand = DefRandom(info, name, loc.tpe, None)
+      Block(List(rand, Connect(info, loc, Reference(rand))))
+    case other =>
+      val info = other match {
+        case h: HasInfo => h.info
+        case _ => NoInfo
+      }
+      val prefix = other match {
+        case c: Connect => c.loc.serialize.replace('.', '_')
+        case h: HasName => h.name
+        case _ => ""
+      }
+      val rands = mutable.ListBuffer[Statement]()
+      val stmt = other.mapExpr(onExpr(namespace, opt, prefix, info, rands, _)).mapStmt(onStmt(namespace, opt, _))
+      if (rands.isEmpty) { stmt }
+      else { Block(rands.toList :+ stmt) }
+  }
+
+  private def onExpr(
+    namespace: Namespace,
+    opt:       InvalidToRandomOptions,
+    prefix:    String,
+    info:      Info,
+    rands:     mutable.ListBuffer[Statement],
+    e:         Expression
+  ): Expression =
+    e.mapExpr(onExpr(namespace, opt, prefix, info, rands, _)) match {
+      case ValidIf(_, value, tpe) if tpe == ClockType =>
+        // we currently assume that clock are always valid
+        // TODO: is that a good assumption?
+        value
+      case ValidIf(cond, value, tpe) if opt.randomizeInvalidSignals =>
+        makeRand(namespace, prefix, info, rands, cond, tpe, value, invert = true)
+      case d @ DoPrim(PrimOps.Div, Seq(_, den), _, tpe) if opt.randomizeDivisionByZero =>
+        val denIsZero = Utils.eq(den, Utils.getGroundZero(den.tpe.asInstanceOf[GroundType]))
+        makeRand(namespace, prefix, info, rands, denIsZero, tpe, d, invert = false)
+      case other => other
+    }
+
+  private def makeRand(
+    namespace: Namespace,
+    prefix:    String,
+    info:      Info,
+    rands:     mutable.ListBuffer[Statement],
+    cond:      Expression,
+    tpe:       Type,
+    value:     Expression,
+    invert:    Boolean
+  ): Expression = {
+    val name = namespace.newName(if (prefix.isEmpty) "invalid" else prefix + "_invalid")
```

Comment:
```suggestion
  private def makeRand(
    namespace: Namespace,
    prefix:    String,
    info:      Info,
    rands:     mutable.ListBuffer[Statement]
  )(
    cond:      Expression,
    value:     Expression,
    invert:    Boolean
  ): Expression = {
    val tpe = value.tpe
    val name = namespace.newName(if (prefix.isEmpty) "invalid" else prefix + "_invalid")
```

This is minor (since this is private), but I found the uses of this super hard to read, I think it would help to tweak it to make the values you're actually passing (vs. threading through) stand out, maybe via a separate argument list. Also, the `Type` is unnecessary, we can derive that from `value.tpe`.

Reply:
I moved the namespace, prefix, info and rands into a `ExprCtx` case class (`private` of course!)